File: 07-数据存储/04-搜索引擎/索引排序与提前终止/python/main.py
"""索引排序与提前终止 —— 演示入口。"""

import os
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from sortmodel import (SortField, Sort, LeafReader, IndexWriterConfig,
                       SCORE, DOC, STRING, INT, LONG, DOUBLE, FLOAT,
                       CUSTOM, REWRITEABLE, STRING_VAL, INDEX_SORTER_TYPES,
                       FIELD_SCORE, FIELD_DOC)


def main():
    print("=" * 72)
    print("Lucene Sort / SortField 与索引排序")
    print("=" * 72)

    print("\n[1] 哪些 SortField 能做索引排序？")
    for t in (STRING, INT, LONG, DOUBLE, FLOAT, SCORE, DOC, CUSTOM,
              REWRITEABLE, STRING_VAL):
        f = SortField(None, t) if t in (SCORE, DOC) else SortField("f", t)
        print("    %-12s getIndexSorter() = %-5s  needsScores() = %s"
              % (t, f.get_index_sorter(), f.needs_scores()))
    print("    可索引排序的类型共 %d 种：%s"
          % (len(INDEX_SORTER_TYPES), sorted(INDEX_SORTER_TYPES)))

    print("\n[2] Sort 的两个常量")
    print("    Sort.RELEVANCE  = %s  needsScores=%s"
          % (Sort.RELEVANCE, Sort.RELEVANCE.needs_scores()))
    print("    Sort.INDEXORDER = %s  needsScores=%s"
          % (Sort.INDEXORDER, Sort.INDEXORDER.needs_scores()))

    print("\n[3] setIndexSort：有一个字段不可用就整体拒绝")
    cfg = IndexWriterConfig()
    for name, sort in (("INDEXORDER(DOC)", Sort.INDEXORDER),
                       ("RELEVANCE(SCORE)", Sort.RELEVANCE),
                       ("[INT, SCORE]", Sort(SortField("a", INT), FIELD_SCORE)),
                       ("[INT, STRING]", Sort(SortField("a", INT), SortField("b", STRING)))):
        try:
            cfg.set_index_sort(sort)
            print("    %-18s → 接受，indexSortFields=%s" % (name, sorted(cfg.index_sort_fields)))
        except ValueError as e:
            print("    %-18s → 拒绝：%s" % (name, e))

    print("\n[4] getPrimarySortField：段上真正生效的主排序字段")
    sf_a, sf_b = SortField("a", INT), SortField("b", STRING)
    sort_ab = Sort(sf_a, sf_b)
    cases = [
        ("正常", LeafReader(100, ["a", "b"], {"sort": sort_ab})),
        ("a 在本段无值", LeafReader(100, ["b"], {"sort": sort_ab})),
        ("a、b 都无值", LeafReader(100, [], {"sort": sort_ab})),
        ("a 全段同值", LeafReader(100, ["a", "b"], {"sort": sort_ab},
                                {"a": {"doc_count": 100, "min": 7, "max": 7}})),
        ("a 值有差异", LeafReader(100, ["a", "b"], {"sort": sort_ab},
                                {"a": {"doc_count": 100, "min": 1, "max": 9}})),
        ("skipper 只覆盖 50/100", LeafReader(100, ["a", "b"], {"sort": sort_ab},
                                        {"a": {"doc_count": 50, "min": 7, "max": 7}})),
        ("两个字段都被跳过", LeafReader(100, ["a", "b"], {"sort": sort_ab},
                               {"a": {"doc_count": 100, "min": 0, "max": 0},
                                "b": {"doc_count": 100, "min": 3, "max": 3}})),
        ("段无索引排序", LeafReader(100, ["a", "b"], {"sort": None})),
    ]
    for label, reader in cases:
        p = Sort.get_primary_sort_field(reader)
        print("    %-22s → %s" % (label, p.get_field() if p else "None"))

    print("\n[5] IndexWriterConfig 默认值")
    print("    DEFAULT_MAX_FULL_FLUSH_MERGE_WAIT_MILLIS = %d"
          % IndexWriterConfig.DEFAULT_MAX_FULL_FLUSH_MERGE_WAIT_MILLIS)
    print("    DEFAULT_RAM_BUFFER_SIZE_MB = %.1f" % IndexWriterConfig.DEFAULT_RAM_BUFFER_SIZE_MB)
    print("    DEFAULT_MAX_BUFFERED_DOCS = %d（DISABLE_AUTO_FLUSH）"
          % IndexWriterConfig.DEFAULT_MAX_BUFFERED_DOCS)
    c = IndexWriterConfig()
    print("    默认 full-flush 合并开启 = %s；设为 0 后 = %s"
          % (c.full_flush_merge_enabled,
             c.set_max_full_flush_merge_wait_millis(0).full_flush_merge_enabled))

    print("\n" + "=" * 72)


if __name__ == "__main__":
    main()

File: 07-数据存储/04-搜索引擎/索引排序与提前终止/go/main.go
// 索引排序与提前终止 —— Go 侧演示入口（与 python/main.py 同题）。
package main

import "fmt"

func main() {
	fmt.Println("[1] 哪些 SortField 能做索引排序？")
	for _, t := range []string{TypeString, TypeInt, TypeLong, TypeDouble, TypeFloat,
		TypeScore, TypeDoc, TypeCustom, TypeRewritable, TypeStringVal} {
		field := "f"
		if t == TypeScore || t == TypeDoc {
			field = ""
		}
		sf, _ := newSortField(field, t, false, nil)
		fmt.Printf("    %-12s getIndexSorter()=%-5v needsScores()=%v\n",
			t, sf.GetIndexSorter(), sf.NeedsScores())
	}

	fmt.Println("\n[2] Sort 的两个常量")
	fmt.Printf("    RELEVANCE  = %s needsScores=%v\n", SortRelevance, SortRelevance.NeedsScores())
	fmt.Printf("    INDEXORDER = %s needsScores=%v\n", SortIndexOrder, SortIndexOrder.NeedsScores())

	fmt.Println("\n[3] SetIndexSort：有一个字段不可用就整体拒绝")
	cfg := NewIndexWriterConfig()
	ia, _ := newSortField("a", TypeInt, false, nil)
	ib, _ := newSortField("b", TypeString, false, nil)
	for _, c := range []struct {
		name string
		s    *Sort
	}{{"INDEXORDER(DOC)", SortIndexOrder}, {"RELEVANCE(SCORE)", SortRelevance},
		{"[INT, SCORE]", NewSort(ia, FieldScore)}, {"[INT, STRING]", NewSort(ia, ib)}} {
		if _, err := cfg.SetIndexSort(c.s); err != nil {
			fmt.Printf("    %-18s -> 拒绝：%v\n", c.name, err)
		} else {
			fmt.Printf("    %-18s -> 接受，indexSortFields=%v\n", c.name, cfg.SortedFieldNames())
		}
	}

	fmt.Println("\n[4] GetPrimarySortField：段上真正生效的主排序字段")
	sab := NewSort(ia, ib)
	cases := []struct {
		label string
		r     *LeafReader
	}{
		{"正常", NewLeafReader(100, []string{"a", "b"}, sab, nil)},
		{"a 在本段无值", NewLeafReader(100, []string{"b"}, sab, nil)},
		{"a、b 都无值", NewLeafReader(100, nil, sab, nil)},
		{"a 全段同值", NewLeafReader(100, []string{"a", "b"}, sab,
			map[string]*Skipper{"a": {100, 7, 7}})},
		{"a 值有差异", NewLeafReader(100, []string{"a", "b"}, sab,
			map[string]*Skipper{"a": {100, 1, 9}})},
		{"skipper 只覆盖 50/100", NewLeafReader(100, []string{"a", "b"}, sab,
			map[string]*Skipper{"a": {50, 7, 7}})},
		{"两个字段都被跳过", NewLeafReader(100, []string{"a", "b"}, sab,
			map[string]*Skipper{"a": {100, 0, 0}, "b": {100, 3, 3}})},
		{"段无索引排序", NewLeafReader(100, []string{"a", "b"}, nil, nil)},
	}
	for _, c := range cases {
		p := GetPrimarySortField(c.r)
		name := "None"
		if p != nil {
			name = p.GetField()
			if name == "" {
				name = p.String()
			}
		}
		fmt.Printf("    %-22s -> %s\n", c.label, name)
	}

	fmt.Println("\n[5] IndexWriterConfig 默认值")
	fmt.Printf("    DEFAULT_MAX_FULL_FLUSH_MERGE_WAIT_MILLIS=%d\n",
		DefaultMaxFullFlushMergeWaitMillis)
	fmt.Printf("    DEFAULT_RAM_BUFFER_SIZE_MB=%.1f  DEFAULT_MAX_BUFFERED_DOCS=%d\n",
		DefaultRamBufferSizeMB, DefaultMaxBufferedDocs)
	c := NewIndexWriterConfig()
	before := c.FullFlushMergeEnabled()
	c.MaxFullFlushMergeWaitMillis = 0
	fmt.Printf("    full-flush 合并：默认=%v，设为 0 后=%v\n", before, c.FullFlushMergeEnabled())
}
